internal/batching/batching.go

```diff
@@ -0,0 +1,272 @@
+	return func(b *Batcher) error {
+		b.log = log
+		return nil
+	}
+}
+
+// WithBatchSize sets the batch size at which to emit at. So if you set this to 1000, it will
+// emit when it has 1000 items in the batch if we haven't hit the timespan. If the timespan
+// is hit, it will emit regardless of the batch size.
+func WithBatchSize(size int) Option {
+	return func(b *Batcher) error {
+		if size < 0 {
+			return errors.New("batch size must be greater than 0")
+		}
+		b.batchSize = size
+		return nil
+	}
+}
+
+// New creates a new Batcher.
+func New(ctx context.Context, in <-chan data.Entry, out chan Batches, timespan time.Duration, options ...Option) (*Batcher, error) {
+	if in == nil || out == nil {
+		return nil, errors.New("can't call Batcher.New() with a nil in or out channel")
+	}
+
+	b := &Batcher{
+		timespan: timespan,
+		in:       in,
+		out:      out,
+		log:      slog.Default(),
+	}
+	b.current = getBatches()
+	b.emitter = b.emit
+
+	for _, o := range options {
+		if err := o(b); err != nil {
+			return nil, err
+		}
+	}
+
+	go b.run()
+
+	return b, nil
+}
+
+// run runs the Batcher loop.
+func (b *Batcher) run() {
+	defer close(b.out)
+
+	timer := time.NewTimer(b.timespan)
+	ticker := time.NewTicker(b.timespan)
+	defer ticker.Stop()
+
+	for {
+		timer.Reset(b.timespan)
+		exit, err := b.handleInput(timer.C)
+		if err != nil {
+			b.log.Error(err.Error())
+		}
+		if exit {
+			return
+		}
+	}
+}
+
+// handleInput handles the input data and batching when the ticker fires.
+func (b *Batcher) handleInput(tick <-chan time.Time) (exit bool, err error) {
+	select {
+	case data, ok := <-b.in:
+		if !ok {
+			return true, nil
+		}
+		if err := b.handleData(data); err != nil {
+			return false, err
+		}
+		if b.batchSize > 0 && len(b.current) == b.batchSize {
+			b.emitter()
+		}
+	case <-tick:
+		if len(b.current) == 0 {
+			return false, nil
+		}
+		b.emitter()
+	}
+	return false, nil
+}
+
+// emit emits the current batches and preps for the new batches. This is assigned
+// to b.emitter by New() at runtime.
+func (b *Batcher) emit() {
+	batches := b.current
+	n := getBatches()
+	b.current = n
+	b.out <- batches
+}
+
+type getMeta interface {
+	GetCreationTimestamp() metav1.Time
+}
+
+// handleData handles putting the data into the current batch.
+func (b *Batcher) handleData(entry data.Entry) error {
+	batch, ok := b.current[entry.SourceType()]
+	if !ok {
+		batch = getBatch()
+	}
+
+	if entry.UID() == "" {
+		return errors.New("no UID for entry")
+	}
+
+	if entry.ChangeType() == data.CTDelete {
+		batch.Map()[entry.UID()] = entry
+		b.current[entry.SourceType()] = batch
+		return nil
+	}
+	old, ok := batch.Map()[entry.UID()]
+	if !ok {
+		batch.Map()[entry.UID()] = entry
+		b.current[entry.SourceType()] = batch
+		return nil
+	}
+	ots := old.Object().(getMeta).GetCreationTimestamp().Time
+	nts := entry.Object().(getMeta).GetCreationTimestamp().Time
+	if nts.After(ots) {
+		batch.Map()[entry.UID()] = entry
```

Comment:
Considering this is for auditing, I'm wondering if it's an issue if we're throwing out changes that haven't been sent to ARG yet.

Though that could lead to a larger buildup of unsent changes when there's already a problem.

Reply:
Part of the initial design agreements was with Jim was to remove older states in batches.  If later they want to change that parameter, certainly makes it easier on us.